src/mame/machine/archimedes_keyb.cpp

```diff
@@ -0,0 +1,392 @@
+//-------------------------------------------------
+
+void archimedes_keyboard_device::device_add_mconfig(machine_config &config)
+{
+	I8051(config, m_mcu, 12_MHz_XTAL);  // Philips 8051AH-2
+	m_mcu->port_in_cb<0>().set([this]() { return m_keyboard[m_mux & 0x0f]->read() & 0xff; });
+	m_mcu->port_in_cb<1>().set([this]() { return (m_keyboard[m_mux & 0x0f]->read() >> 8) & 0xff; });
+	m_mcu->port_in_cb<2>().set(FUNC(archimedes_keyboard_device::mouse_r));
+	m_mcu->port_in_cb<3>().set([this]() { return m_mouse[2]->read() & 0xff; });
+	m_mcu->port_out_cb<2>().set([this](uint8_t data) { m_mux = data; });
+	m_mcu->port_out_cb<3>().set(FUNC(archimedes_keyboard_device::leds_w));
+	m_mcu->serial_rx_cb().set([this]() { return get_received_char(); });
+	m_mcu->serial_tx_cb().set(FUNC(archimedes_keyboard_device::tx_w));
+}
+
+//-------------------------------------------------
+//  rom_region - device-specific ROM region
+//-------------------------------------------------
+
+ROM_START( archimedes_keyboard )
+	ROM_REGION( 0x1000, "mcu", 0 )
+	ROM_LOAD("acorn_0280,022-01_philips_8051ah-2.bin", 0x0000, 0x1000, CRC(95095d38) SHA1(b9fd0000d77987f76fd48348fe2c6988818f40cb))
+ROM_END
+
+
+const tiny_rom_entry *archimedes_keyboard_device::device_rom_region() const
+{
+	return ROM_NAME( archimedes_keyboard );
+}
+
+void archimedes_keyboard_device::tra_complete()
+{
+}
+
+void archimedes_keyboard_device::rcv_complete()
+{
+	receive_register_extract();
+	m_mcu->set_input_line(MCS51_RX_LINE, ASSERT_LINE);
+	m_mcu->set_input_line(MCS51_RX_LINE, CLEAR_LINE);
+}
+
+void archimedes_keyboard_device::tra_callback()
+{
+	m_kout(transmit_register_get_data_bit());
+}
+
+void archimedes_keyboard_device::device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr)
+{
+	//  Mouse quadrature signals
+	//  Increase      Decrease
+	//  REF  DIR      REF  DIR
+	//   1    1        1    1
+	//   1    0        0    1
+	//   0    0        0    0
+	//   0    1        1    0
+
+	switch (m_mouse_phase)
+	{
+	case 0:
+		m_mouse_xdir = m_mouse_xref = ASSERT_LINE;
+		m_mouse_ydir = m_mouse_yref = ASSERT_LINE;
+		break;
+	case 1:
+		m_mouse_xref = m_mouse_x < 0 ? CLEAR_LINE : ASSERT_LINE;
+		m_mouse_xdir = m_mouse_x > 0 ? CLEAR_LINE : ASSERT_LINE;
+		m_mouse_yref = m_mouse_y < 0 ? CLEAR_LINE : ASSERT_LINE;
+		m_mouse_ydir = m_mouse_y > 0 ? CLEAR_LINE : ASSERT_LINE;
+		break;
+	case 2:
+		m_mouse_xdir = m_mouse_xref = m_mouse_x != 0 ? CLEAR_LINE : ASSERT_LINE;
+		m_mouse_ydir = m_mouse_yref = m_mouse_y != 0 ? CLEAR_LINE : ASSERT_LINE;
+		break;
+	case 3:
+		m_mouse_xref = m_mouse_x > 0 ? CLEAR_LINE : ASSERT_LINE;
+		m_mouse_xdir = m_mouse_x < 0 ? CLEAR_LINE : ASSERT_LINE;
+		m_mouse_yref = m_mouse_y > 0 ? CLEAR_LINE : ASSERT_LINE;
+		m_mouse_ydir = m_mouse_y < 0 ? CLEAR_LINE : ASSERT_LINE;
+		if (m_mouse_x > 0)    m_mouse_x--;
+		if (m_mouse_x < 0)    m_mouse_x++;
+		if (m_mouse_y > 0)    m_mouse_y--;
+		if (m_mouse_y < 0)    m_mouse_y++;
+		break;
```

Comment:
You shouldn’t be using the `_LINE` constant where you just want bits.  They’re only for line inputs that support magic pulse/hold values.